crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -201,13 +201,54 @@ impl SettlementEncoder {
                 )?
             }
             OrderClass::Limit => {
-                todo!()
+                // Limit orders are fill-or-kill so we have to execute the total amount.
+                let executed_amount = order.data.sell_amount + order.data.fee_amount;
+                let buy_price = self.custom_price_for_limit_order(&order)?;
+
+                self.add_custom_price_trade(
+                    order,
+                    executed_amount,
+                    scaled_unsubsidized_fee,
+                    buy_price,
+                )?
             }
         };
         self.pre_interactions.extend(interactions.pre.into_iter());
         Ok(execution)
     }
 
+    /// Uses the uniform clearing prices to compute the individual buy token price to satisfy the
+    /// original limit order with was adjusted to account for the `surplus_fee` (see
+    /// `build_synthetic_order_if_limit_order()`).
+    /// Returns an error if the UCP doesn't contain the `sell_token` or if under- or overflows
+    /// happen during the computation.
+    fn custom_price_for_limit_order(&self, order: &Order) -> Result<U256> {
+        // This is the equation the solver tried to satisfy:
+        // uniform_sell_price * sell_amount = uniform_buy_price * buy_amount
+        //
+        // And this is what the solution actually has to satisfy:
+        // uniform_sell_price * (sell_amount + surplus_fee) = custom_buy_price * buy_amount
+        // That means we can solve for the custom_buy_price like this:
+        // custom_buy_price = (uniform_sell_price * (sell_amount + surplus_fee)) / buy_amount
+
+        let uniform_sell_price = self
+            .clearing_prices()
+            .get(&order.data.sell_token)
+            .context("missing sell token price")?;
+
+        let sell_amount_with_surplus_fees = order
+            .data
+            .sell_amount
+            .checked_add(order.metadata.surplus_fee)
+            .context("adjusting for surplus_fee would overflow sell_amount")?;
+
+        sell_amount_with_surplus_fees
+            .checked_mul(*uniform_sell_price)
+            .context("buy_price calculation failed")?
+            .checked_ceil_div(&order.data.buy_amount)
```

Comment:
Still have to figure out if this has to be `ceil_div()` or just `div()`.

Reply:
🤷 

Joking aside, I _think_ it should be `div` to favor the user (round down `buy_token_price` means buy token is worth less and the user gets more) and `ceil_div` to favor the solver.

Reply:
My intuition is that we should favor the user here, but I haven't fully thought it through yet.